### PR TITLE
feat: added resource deletion

### PR DESCRIPTION
**Pull request by GugaGongadze**
## Description of change

Added service resources deletion.

## How has this been tested? (if applicable)

N/A

### CODE REVIEW

File: resource-recorder/src/dal.rs [hunk 1]
```diff
@@ -177,6 +180,18 @@ impl Dal for Sqlite {
...
+        sqlx::query("DELETE FROM resources WHERE project_id = ? AND service_id = ? AND type = ?")
+            .bind(resource.project_id.map(|u| u.to_string()))
+            .bind(resource.service_id.map(|u| u.to_string()))
+            .bind(resource.r#type)
+            .execute(&self.pool)
+            .await
+            .map(|_| ())?;
+
+        Ok(())
+    }
```
Comment (oddgrd) [inline resource-recorder/src/dal.rs:193]: I think you missed a copy-paste here, we need to rewrite this to rather get a specific resource.

File: deployer/src/deployment/run.rs [hunk 2]
```diff
@@ -574,20 +590,26 @@ mod tests {
...
+            _service_id: &ulid::Ulid,
+            _type: ResourceType,
+            _claim: Claim,
+        ) -> Result<ResourceResponse, Self::Err> {
+            Ok(ResourceResponse {
+                success: true,
+                message: "dummy impl".to_string(),
+                resource: None,
+            })
         }
```
Comment (oddgrd) [inline deployer/src/deployment/run.rs:605]: We need to refactor this to match the type from the trait

File: deployer/src/handlers/mod.rs [hunk 3]
```diff
@@ -320,6 +333,60 @@ pub async fn get_service_resources(
...
+        .ok_or_else(|| Error::NotFound("resource not found".to_string()))?;
+
+    // resource_recorder_client
+    // resource_manager
+    //     .delete_resource(&project_name, &resource_type)
+    //     .await
+    //     .map_err(|e| Error::Custom(e.into()))?;
+    persistence
+        .delete_resource(&service.id, persistence_resource_type, claim)
+        .await?;
```
Comment (oddgrd) [inline deployer/src/handlers/mod.rs:385]: I'm thinking we can try to roll the provisioner resource deletion in `persistence.delete_resource`.

File: deployer/src/main.rs [hunk 4]
```diff
@@ -72,6 +76,10 @@ async fn main() {
         Backend::Deployer,
     );
 
+    let resource_manager = ResourceManager::new(&args.provisioner_address)
+        .await
+        .unwrap();
```
Comment (oddgrd) [inline deployer/src/main.rs:81]: I think we can remove this ResourceManager struct, and rather pass the provisioner address to persistence, like we do with resource recorder at the start of this function (main). Then we put the provisioner client in the persistence struct, like we do for the resource manager client.

File: deployer/src/persistence/mod.rs [hunk 5]
```diff
@@ -466,6 +466,35 @@ impl ResourceManager for Persistence {
...
+        claim: Claim,
+    ) -> Result<()> {
+        sqlx::query(r#"DELETE FROM resources WHERE service_id = ? AND type = ?"#)
+            .bind(service_id.to_string())
+            .bind(r#type)
+            .execute(&self.pool)
+            .await
+            .map(|_| ())
+            .map_err(Error::from)
+    }
```
Comment (oddgrd) [inline deployer/src/persistence/mod.rs:497]: We shouldn't need to delete from the sqlite database, resource deletion will only be available to new deployers, and they will use resource-recorder to record resources, not the sqlite database. So get_resource should use the resource-recorder client. Delete resource should first use the provisioner client to delete the resource, then if that succeeds it should use the resource-recorder client to delete the resource from the resource-recorder state.

**Review by oddgrd — COMMENTED** [review]
Thanks Guga, this is getting there! I left some suggestions for how to solve some of the errors, as well as some suggestions for how we might re-structure this in deployer.

File: deployer/src/persistence/mod.rs (hunk 5 above)
Comment (oddgrd) [reply deployer/src/persistence/mod.rs:497]: Hmm, we might also need to check that the resource is something the provisioner should delete. The provisioner only deals with databases currently. But we can probably start with the above comment and then tweak it after.

File: deployer/src/handlers/mod.rs (hunk 3 above)
Comment (iulianbarbu) [reply deployer/src/handlers/mod.rs]: I think this should be fixed here: https://github.com/shuttle-hq/shuttle/pull/1256/commits/61eb0ebf6b1914f55ecc97e80a4676824ded85bf. LMK if you meant something else.

File: deployer/src/persistence/mod.rs (hunk 5 above)
Comment (iulianbarbu) [reply deployer/src/persistence/mod.rs]: I think this commit reflects some of your suggestions: https://github.com/shuttle-hq/shuttle/pull/1256#discussion_r1338706181.

File: deployer/src/main.rs (hunk 4 above)
Comment (iulianbarbu) [reply deployer/src/main.rs]: Looking into this.

Comment (iulianbarbu) [reply deployer/src/main.rs]: This should be covered in the latest changes.

File: deployer/src/handlers/mod.rs [hunk 6]
```diff
@@ -157,6 +169,7 @@ impl RouterBuilder {
             )
             .layer(Extension(persistence))
             .layer(Extension(deployment_manager))
+            // .layer(Extension(resource_manager))
```
Comment (chesedo) [inline deployer/src/handlers/mod.rs:172]: Comment

File: deployer/src/persistence/mod.rs [hunk 7]
```diff
@@ -459,13 +496,79 @@ impl ResourceManager for Persistence {
...
+            .await
+            .map_err(PersistenceError::ResourceRecorder)
+            .map(|res| res.into_inner());
+    }
+
+    async fn delete_resource(
+        &mut self,
+        project_name: String,
+        service_id: &Ulid,
+        resource_type: String,
```
Comment (chesedo) [inline deployer/src/persistence/mod.rs:535]: Shouldn't this type be the resource type (and not a string)?

File: deployer/src/persistence/mod.rs [hunk 8]
```diff
@@ -459,13 +496,79 @@ impl ResourceManager for Persistence {
...
             }
         }
 
         Ok(res)
     }
+
+    async fn get_resource(
+        &mut self,
+        service_id: &Ulid,
+        r#type: String,
```
Comment (chesedo) [inline deployer/src/persistence/mod.rs:510]: Same here, this should be the resource enum to reduce errors

**Review by chesedo — COMMENTED** [review]
Just minor notes

**Review by jonaro00 — COMMENTED** [review]
Why are we deleting resources by type, and not by id?

File: deployer/src/persistence/mod.rs (hunk 8 above)
Comment (iulianbarbu) [reply deployer/src/persistence/mod.rs]: Done.

File: deployer/src/handlers/mod.rs (hunk 6 above)
Comment (iulianbarbu) [reply deployer/src/handlers/mod.rs]: Done.

**Review by oddgrd — APPROVED** [review]

**Pull request merged**